src/calendardata.h

```diff
@@ -90,6 +205,200 @@ struct Event {
+                    return CalendarEvent::RecurMonthlyByLastDayOfWeek;
+                }
+            }
+        } else if (rt == KCalendarCore::Recurrence::rYearlyMonth && freq == 1) {
+            return CalendarEvent::RecurYearly;
+        }
+
+        return CalendarEvent::RecurCustom;
+    }
+
+    void toKRecurrence(KCalendarCore::Event &event) const
+    {
+        CalendarEvent::Recur oldRecur = fromKRecurrence(event);
+
+        if (recur == CalendarEvent::RecurOnce)
+            event.recurrence()->clear();
+
+        if (oldRecur != recur
+            || recur == CalendarEvent::RecurMonthlyByDayOfWeek
+            || recur == CalendarEvent::RecurMonthlyByLastDayOfWeek
+            || recur == CalendarEvent::RecurWeeklyByDays) {
+            switch (recur) {
+            case CalendarEvent::RecurOnce:
+                break;
+            case CalendarEvent::RecurDaily:
+                event.recurrence()->setDaily(1);
+                break;
+            case CalendarEvent::RecurWeekly:
+                event.recurrence()->setWeekly(1);
+                break;
+            case CalendarEvent::RecurBiweekly:
+                event.recurrence()->setWeekly(2);
+                break;
+            case CalendarEvent::RecurWeeklyByDays: {
+                QBitArray rDays(7);
+                rDays.setBit(0, recurWeeklyDays & CalendarEvent::Monday);
+                rDays.setBit(1, recurWeeklyDays & CalendarEvent::Tuesday);
+                rDays.setBit(2, recurWeeklyDays & CalendarEvent::Wednesday);
+                rDays.setBit(3, recurWeeklyDays & CalendarEvent::Thursday);
+                rDays.setBit(4, recurWeeklyDays & CalendarEvent::Friday);
+                rDays.setBit(5, recurWeeklyDays & CalendarEvent::Saturday);
+                rDays.setBit(6, recurWeeklyDays & CalendarEvent::Sunday);
+                event.recurrence()->setWeekly(1, rDays);
+                break;
+            }
+            case CalendarEvent::RecurMonthly:
+                event.recurrence()->setMonthly(1);
+                break;
+            case CalendarEvent::RecurMonthlyByDayOfWeek: {
+                event.recurrence()->setMonthly(1);
+                const QDate at(event.dtStart().date());
+                event.recurrence()->addMonthlyPos((at.day() - 1) / 7 + 1, at.dayOfWeek());
+                break;
+            }
+            case CalendarEvent::RecurMonthlyByLastDayOfWeek: {
+                event.recurrence()->setMonthly(1);
+                const QDate at(event.dtStart().date());
+                event.recurrence()->addMonthlyPos(-1, at.dayOfWeek());
+                break;
+            }
+            case CalendarEvent::RecurYearly:
+                event.recurrence()->setYearly(1);
+                break;
+            case CalendarEvent::RecurCustom:
+                // Unable to handle the recurrence rules, keep the existing ones.
+                break;
+            }
+        }
+    }
```

Comment:
Earlier this was mostly data type definition file now there's a bunch of code too. Could create a .cpp file for the methods?

Reply:
Hm, could it be a middle ground if the code would reside now in calendarutils.cpp instead of a separate file? That used to have parts of it already earlier.

Reply:
Sure, I've moved all the `Event`code into `calendarutils.cpp`.